pandas/core/arrays/datetimelike.py

```diff
@@ -479,12 +481,13 @@ def __setitem__(
                 raise ValueError("setting an array element with a sequence.")
 
             if (not is_slice
-                    and len(key) != len(value)
+                    and len(cast(Sequence, key)) != len(value)
```

Comment:
What does the cast do here?

Reply:
Yea I suppose this is similar to the question above. Would be nice if `is_list_like` could signal to the type checker that we are working with a sequence

Reply:
@gwrome can you just do `key = cast(Sequence, key)` once at the beginning of this branch instead?

Reply:
This will work, but I have reservations:
```py
if is_list_like(value):
    is_slice = isinstance(key, slice)

    if lib.is_scalar(key):
        raise ValueError("setting an array element with a sequence.")

    key = cast(Sequence, key)  # <-- <A>
    if (not is_slice
            and len(key) != len(value)
            and not com.is_bool_indexer(key)):
           # <-- <B>
        msg = ("shape mismatch: value array of length '{}' does not "
                "match indexing result of length '{}'.")
        raise ValueError(msg.format(
            len(key), len(value)))
    if not is_slice and len(key) == 0:
        return
```
If we put `key = cast(Sequence, key)` at `<A>`, we're not guaranteed that `key` is a `Sequence` and not a `slice`. We only know that key is not scalar at that point. I presume that's why both following conditions start with `not is_slice`. If we put the cast at `<B>` it only saves us the one cast in the ValueError message and doesn't affect the casts in the two conditions.

Factoring the slice check out of both if statements works for us, and lets us guarantee that `key` is a `Sequence` when we cast. But it's a more substantial change than I would normally be comfortable making as part of cleaning up type annotation errors:
```py
if not is_slice:
    key = cast(Sequence, key)
    if (len(key) != len(value)
            and not com.is_bool_indexer(key)):
        msg = ("shape mismatch: value array of length '{}' does "
                "not match indexing result of length '{}'.")
        raise ValueError(msg.format(
            len(key), len(value)))
    if len(key) == 0:
        return
```

Reply:
I feel like an old man yelling "get off my lawn", but this adds cognitive overhead for me since I have to go learn what `cast` and `Sequence` are.  When typing is in comments and signatures I can ignore it, but not in places like this.  So again: does this gain us anything other than making the code clearer to mypy?  Because of the cost of "making code clearer to mypy" is "making code less clear to readers [me]", that isn't an obvious win.

Reply:
@jbrockmendel it's certainly a legitimate concern. As you might have seen from the link in https://github.com/pandas-dev/pandas/pull/26071#discussion_r275522392 this is something which arguably could be more robustly handled by the type checker, but that feature doesn't yet exist. 

The flip side though is that this is part of the standard library, so I don't think we are doing anything too farfetched that isn't aligned with the evolution of the language. Do you have an alternative in mind or is it a general hesitation with type checking at this point?

Reply:
@WillAyd I just pushed a commit with the second option in it. Does that address your concerns?